src/url-params.ts

```diff
@@ -38,12 +44,24 @@ export class AnalyticsUrlParams {
 		return list.length > 0 ? list[0] : null;
 	}
 
+	private setSessionId(storedValue: string | null | undefined) {
+		if (storedValue) {
```

Comment:
I have a discussion point: Shall we follow Amplitude's logic and define timeout windows? 
[Relevant Amplitude docs](https://help.amplitude.com/hc/en-us/articles/115002323627-Tracking-Sessions#h_c6714eea-04a4-48ce-b2e0-e0379213ad95)

When the incoming `sessionId` is from 1 hour ago or longer, do we still want to apply it as the session of the new events?

Reply:
I suggest applying the incoming `sessionId` value only if it's in an acceptable time window, like 30 minutes or perhaps 1 hour.

Reply:
It should be 30minutes ~ 1hour without any activity, if the users spends 4 hours on the dashboard by actively doing something, then that should be the same session.

Reply:
@sradevski right. Then we might need to record and check the timestamp of the navigation moment. `sessionId` is essentially the epoch timestamp of the first event in a session. All the following events in the same session use the same timestamp. So we could decide not to modify `sessionId` when a user was inactive for X minutes after navigating to the next component. 

I'm thinking particularly the case of users saving / sharing URLs with `sessionId` param. Though I'm not sure how large of an impact these will have, it'd be nice not to pollute our session data.

Reply:
I am familiar with the technique of splitting sessions after a certain amount of inactivity and used 30 min timeouts in previous projects.

If done by ourselves, it certainly means keeping track of the *latest* activity timestamp to decide if the subsequent event qualifies for the same session.
But quickly looking at the docs ([this](https://help.amplitude.com/hc/en-us/articles/115002323627-Tracking-Sessions#h_7a795c15-e644-4cec-9f99-54dc24972f27) section specifically) my understanding is that this should be handled by Amplitude automatically. If that's the case, I'd be hesitant to re-implement the same logic (a.k.a. reinvent the wheel).

Let me know if I'm missing any crucial aspects here 🤔

Reply:
Sounds like it's best to test the default behaviour here before we start adding extra logic. I don't know if / how Amplitude would behave when we apply a custom `sessionId`.

Reply:
Good idea!

Reply:
@pranasziaukas I'm running both `analytics-client` and `analytics-backend` locally to observe Amplitude's behaviour around custom `sessionId`. Triggering from `/example/main.js` and using the default values, I could see my events collected under `Stg - test` project (specifically [this user](https://analytics.amplitude.com/balena/project/252097/search/amplitude_id%3D148200232010)). First of all, I suggest you to run the full stack locally yourself as well. I found it quite helpful. Moving on this `sessionId` discussion: seems like Amplitude happily merges all the events sent under the same `sessionId`. There is no extra logic based on timeout windows. 

[The top events](https://analytics.amplitude.com/balena/project/252097/search/amplitude_id%3D148200232010?sessionHandle=XDtXqKI_XeCIie6_CKBbBRK_9jB_0000-0000-0000&eventId=6d5fb40a-6a1d-11eb-8433-0a0cabafa3f1) on February 8 are merged to the session from March 18 (`session id = 1584530367112`). Next using `session id = 1612345671145`, I merged [today's events](https://analytics.amplitude.com/balena/project/252097/search/amplitude_id%3D148200232010?sessionHandle=XdnSn3p_XdnSn36_DXEG0b2_9jB_7db36c83-5eb7-4f56-a377-4d2a0923fac4R&eventId=20a0da18-6a1d-11eb-96b7-063a4d28002b) with the events from last week. 

So either on `analytics-client` or on `analytics-backend` it might help to introduce some extra logic ourselves to prevent such "session pollution". Would you agree?

Reply:
Thanks for the instructions Firat, I'll give it a go tomorrow.

> So either on `analytics-client` or on `analytics-backend` it might help to introduce some extra logic ourselves to prevent such "session pollution". Would you agree?

I certainly agree that a sane session splitting between prolonged times of inactivity is very much useful and needed. It's just a bit disappointing if Amplitude's client cannot initiate a new session automatically. In that case, I'd at least like to better understand where this limitation is coming from as the documentation (in your first post above) describes default timeout windows of 30 mins, so we were not just speculating wildly.

Reply:
@gelbal could you elaborate (even more :)) on running the test flow locally?

I ran into issues while starting the `analytics-backend` https://github.com/balena-io/analytics-backend/issues/26
Furthermore, my impression was that `/example/index.html` could be opened, which in turn loads the `main.js`. Do you do it differently?

Reply:
@pranasziaukas right. I see the same redis errors on my end. I ignored them as they were not relevant to our exercise here. `analytics-backend` uses redis to count and throttle the supervisor events that are triggered in small window of time. Good idea to open that issue so we document the devenv setup clearly. 

Anyhow, indeed I was editing `main.js` to test stuff.

Reply:
Thinking about this further: Shall we use the expiry date of cookies to achieve what we want? 
We could set a hard upper limit based on `sessionId` value and `currentTime`. The downsides of this is such hard limit surely impacts the analysis. Though this might be preferred actually.

Reply:
@gelbal, do you mean setting the hard limit once and thus limiting how long a single session could be? I don't say it would be wrong but surely unusual as typically sessions last for as long as there's a continuous activity.

Or do you mean updating the cookie's TTL every time we fire an event? I have no idea if that would be a good practice, to be honest.

Both add to my hesitation to implement the full session logic ourselves. But maybe I just have to let the idea brew in my head :)

Reply:
> Or do you mean updating the cookie's TTL every time we fire an event?

Indeed, this is what I was thinking Pranas. My reasoning is not to have extra logic in the `analytics-client` code but instead utilize the Cookies expiry date to decide until when a passed session ID is valid. 

I'm curious to hear how you'd like to implement this. If it helps, we could raise it to Kimmo to see if he knows a common / best practice here.

Reply:
In the ideal world, I'd like Amplitude's client to manage as much of this as possible i.e. to generate a new session ID after a certain period of inactivity for us.

An illustration:
- Events are being sent under the session X.
- A period of inactivity takes place.
- Events are being sent under the session Y (updated courtesy of Amplitude).
- At any point, we get the relevant session ID because `AnalyticsUrlParams.getSessionId()` prefers the session ID straight from the client.

If I understood correctly, for some reason session timeouts are not fully working somewhere under the hood.

Updating cookies on every action is more like handling the logic on our side, which might be fine but then it may be worth checking with Kimmo.

How does that sound?

Reply:
Agreed Pranas. I could only guess that Amplitude's reasoning is they want their SDK to be flexible when the client overwrites the `sessionId` -- so they don't apply the default timeout logic. Then perhaps there is a way to customize Amplitude behaviour. 

Can you please raise this point to Kimmo? 
It'd help to send the example sessions I linked here where we see how Amplitude doesn't apply timeout on the overwrites.

Reply:
That's cool @gelbal, I'll message Kimmo to get a better understanding of the matter and let the analytics team know.

Reply:
@gelbal when you sent events to an old sessionId, did you tested using getSessionId() after timeout of a current session? or did you just set the session to be an old one from the USer Stream? Im wondering if the getSessionId() has a built in way to see if the session you are trying to retrieve timed out. This is because I saw that theres a method [here](https://amplitude.github.io/Amplitude-JavaScript/#amplitudeclientisnewsession) called isNewSession so it seems that it has some way to figure that out on its own.

Reply:
Good call @laziob.

`isNewSession` docs actually confused me. It says:
> Returns true if a new session was created during initialization, otherwise false.

We already checked the code and docs together on Wednesday and saw no `sessionId` parameter in the initialization config (like there is one for `deviceId`). Anyhow, looking at Amplitude JS source code, the `if` statement above this line explains when Amplitude sets `newSession` to true -- which is run during initialization indeed.
https://github.com/amplitude/Amplitude-JavaScript/blob/14fc693d56ff59a2a82b0eeadc4e7e6d3104026b/src/amplitude-client.js#L173

```js
if (!this._sessionId || !this._lastEventTime || now - this._lastEventTime > this.options.sessionTimeout)
```

So checking for ``isNewSession` might be the solution indeed. Let's test it.

Reply:
To continue my train of thought: Actually, it'd not help in our case since there won't be `_lastEventTime` on the new component. 

We might as well apply the same logic ourselves on `analytics-client` -- which comes back to using something cookie expiry date to track `_lastEventTime`.

Let's see how Amplitude support responds.